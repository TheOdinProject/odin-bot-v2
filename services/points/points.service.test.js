const axios = require('axios');
const PointsService = require('./points.service');
const {
  generateLeaderData,
  generateLeaderDataWithMarkdown,
} = require('../../botCommands/mockData');

/* eslint-disable */
/* eslint max-classes-per-file: ["error", 2] */

class GuildMembersMock {
  members;
  cache;

  constructor(users) {
    this.members = users;
    this.cache = {
      get: (id) => users.filter((member) => member.discord_id === id)[0],
    };
  }
}

class GuildMock {
  members;
  member;

  constructor(users) {
    this.members = new GuildMembersMock(users);
    this.member = (user) => users.filter((member) => member === user)[0];
  }
}

describe('leaderboard subcommand', () => {
  let limit = null;
  let offset = null;
  let reply = '';
  const interactionMock = {
    options: {
      getSubcommand: () => 'leaderboard',
      getInteger: (string) => {
        if (string === 'limit') {
          return limit;
        } else if (string === 'offset') {
          return offset;
        }
      },
    },
    reply: jest.fn((message) => (reply = message)),
  };

  const setUpAxiosMock = (data) =>
    (axios.get = jest.fn().mockResolvedValue({ data }));

  afterEach(() => {
    limit = null;
    offset = null;
    reply = '';
    axios.get.mockReset();
  });

  it('Returns be the first to earn points message if no user in database', async () => {
    const members = generateLeaderData(0);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Returned users length matches limit', async () => {
    limit = 8;
    const members = generateLeaderData(30);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Limit defaults to 5 if limit not provided', async () => {
    const members = generateLeaderData(30);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Limit defaults to 5 if invalid characters provided', async () => {
    limit = 'sdfsdf';
    const members = generateLeaderData(30);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Limit defaults to 5 if limit provided lower than 1', async () => {
    limit = 0;
    const members = generateLeaderData(25);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Limit defaults to 25 if limit provided is higher than 25', async () => {
    limit = 50;
    const members = generateLeaderData(70);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Limit does not exceed users length', async () => {
    limit = 10;
    const members = generateLeaderData(5);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Members who left are filtered out', async () => {
    const members = generateLeaderData(50);
    const limitMembers = members.slice(20, 40);
    interactionMock.guild = new GuildMock(limitMembers);
    setUpAxiosMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Returns users starting from offset', async () => {
    offset = 5;
    const members = generateLeaderData(50);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Offset defaults to 0 if not provided', async () => {
    const members = generateLeaderData(25);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Offset defaults to 0 if invalid characters provided', async () => {
    offset = 'sdfs';
    const members = generateLeaderData(25);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Offset defaults to 0 if negative value provided', async () => {
    offset = -5;
    const members = generateLeaderData(25);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Offset cannot exceed max users length', async () => {
    offset = 70;
    const members = generateLeaderData(50);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Offset show the last users depending on limit if offset is too high', async () => {
    offset = 56;
    limit = 5;
    const members = generateLeaderData(50);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Return the correct limit, moving offset backward if necessary', async () => {
    offset = 56;
    limit = 10;
    const members = generateLeaderData(60);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('escapes markdown correctly', async () => {
    const members = generateLeaderDataWithMarkdown(5);
    setUpAxiosMock(members);
    interactionMock.guild = new GuildMock(members);

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });
});

describe('user subcommand', () => {
  const apiData = [
    { id: 7, discord_id: '7777', points: 4000 }, // Left Server
    { id: 6, discord_id: '6666', points: 3000 }, // Cat
    { id: 5, discord_id: '5555', points: 2000 }, // Left Server
    { id: 3, discord_id: '3333', points: 1000 }, // Dog
    { id: 2, discord_id: '2222', points: 20 }, // Someone
    { id: 4, discord_id: '4444', points: 15 }, // Tree
    { id: 1, discord_id: '1111', points: 1 }, // NotOdin
    { id: 8, discord_id: '8888', points: 1 }, // Mark *down*
  ];

  const guildMock = new GuildMock([
    { displayName: 'NotOdin', discord_id: '1111' },
    { displayName: 'Someone', discord_id: '2222' },
    { displayName: 'Dog', discord_id: '3333' },
    { displayName: 'Tree', discord_id: '4444' },
    { displayName: 'Cat', discord_id: '6666' },
    { displayName: 'Mark *down*', discord_id: '8888' },
  ]);

  beforeEach(() => {
    axios.get = jest.fn().mockResolvedValue({ data: apiData });
  });

  let user = {}; // created for each test each test
  let reply = '';
  const interactionMock = {
    guild: guildMock,
    options: {
      getSubcommand: () => 'user',
      getUser: () => user,
    },
    reply: jest.fn((message) => (reply = message)),
  };

  afterEach(() => {
    reply = '';
    user = {};
  });

  it('Return correct reply when user is not in database', async () => {
    user = { id: '200', username: 'OldUser' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Return correct rank for user ignoring members that has left', async () => {
    user = { id: '3333', username: 'Dog' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();

    user = { id: '2222', username: 'Someone' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();

    user = { id: '4444', username: 'Tree' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Rank append emoji to first ranked user', async () => {
    user = { id: '6666', username: 'Cat' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Format the points word properly for 1 point', async () => {
    user = { id: '1111', username: 'NotOdin' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Formats the points word properly for 0 points', async () => {
    user = { id: '9292', username: 'OldestUser' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Formats the points word correctly for more than 1 points', async () => {
    user = { id: '3333', username: 'Dog' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });

  it('Escapes markdown', async () => {
    user = { id: '8888', username: 'Mark *down*' };

    await PointsService.handleInteraction(interactionMock);
    expect(axios.get).toHaveBeenCalled();
    expect(reply).toMatchSnapshot();
  });
});
