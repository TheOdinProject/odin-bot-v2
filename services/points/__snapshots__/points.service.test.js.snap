// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaderboard subcommand Limit defaults to 5 if invalid characters provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Limit defaults to 5 if limit not provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Limit defaults to 5 if limit provided lower than 1 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Limit defaults to 25 if limit provided is higher than 25 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Limit does not exceed users length 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Members who left are filtered out 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user120 [980 points] :tada: 
2 - user121 [979 points] 
3 - user122 [978 points] 
4 - user123 [977 points] 
5 - user124 [976 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Offset cannot exceed max users length 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "46 - user145 [955 points] 
47 - user146 [954 points] 
48 - user147 [953 points] 
49 - user148 [952 points] 
50 - user149 [951 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Offset defaults to 0 if invalid characters provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Offset defaults to 0 if negative value provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Offset defaults to 0 if not provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Offset show the last users depending on limit if offset is too high 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "46 - user145 [955 points] 
47 - user146 [954 points] 
48 - user147 [953 points] 
49 - user148 [952 points] 
50 - user149 [951 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Return the correct limit, moving offset backward if necessary 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "51 - user150 [950 points] 
52 - user151 [949 points] 
53 - user152 [948 points] 
54 - user153 [947 points] 
55 - user154 [946 points] 
56 - user155 [945 points] 
57 - user156 [944 points] 
58 - user157 [943 points] 
59 - user158 [942 points] 
60 - user159 [941 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Returned users length matches limit 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Returns be the first to earn points message if no user in database 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "Be the first to earn a point!",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand Returns users starting from offset 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`leaderboard subcommand escapes markdown correctly 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "description": "1 - user100 \\|\\|test\\|\\| [1000 points] :tada: 
2 - user101 \\*test\\* [999 points] 
3 - user102 \\*\\*test\\*\\* [998 points] 
4 - user103 \\\`test\\\` [997 points] 
5 - user104 \\|\\|test\\|\\| [996 points] 
",
      "title": "TOP Discord points leaderboard",
    },
  ],
}
`;

exports[`user subcommand Escapes markdown 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Mark \\*down\\* has 1 point",
        },
        {
          "name": "Rank",
          "value": "Mark \\*down\\* is ranked number 6",
        },
      ],
      "title": "TOP Discord points for Mark \\*down\\*",
    },
  ],
}
`;

exports[`user subcommand Format the points word properly for 1 point 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "NotOdin has 1 point",
        },
        {
          "name": "Rank",
          "value": "NotOdin is ranked number 5",
        },
      ],
      "title": "TOP Discord points for NotOdin",
    },
  ],
}
`;

exports[`user subcommand Formats the points word correctly for more than 1 points 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Dog has 1000 points",
        },
        {
          "name": "Rank",
          "value": "Dog is ranked number 2",
        },
      ],
      "title": "TOP Discord points for Dog",
    },
  ],
}
`;

exports[`user subcommand Formats the points word properly for 0 points 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "OldestUser has 0 points",
        },
        {
          "name": "Rank",
          "value": "OldestUser currently has no rank in the leaderboard",
        },
      ],
      "title": "TOP Discord points for OldestUser",
    },
  ],
}
`;

exports[`user subcommand Rank append emoji to first ranked user 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Cat has 3000 points",
        },
        {
          "name": "Rank",
          "value": "Cat is ranked number 1 :tada:",
        },
      ],
      "title": "TOP Discord points for Cat",
    },
  ],
}
`;

exports[`user subcommand Return correct rank for user ignoring members that has left 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Dog has 1000 points",
        },
        {
          "name": "Rank",
          "value": "Dog is ranked number 2",
        },
      ],
      "title": "TOP Discord points for Dog",
    },
  ],
}
`;

exports[`user subcommand Return correct rank for user ignoring members that has left 2`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Someone has 20 points",
        },
        {
          "name": "Rank",
          "value": "Someone is ranked number 3",
        },
      ],
      "title": "TOP Discord points for Someone",
    },
  ],
}
`;

exports[`user subcommand Return correct rank for user ignoring members that has left 3`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "Tree has 15 points",
        },
        {
          "name": "Rank",
          "value": "Tree is ranked number 4",
        },
      ],
      "title": "TOP Discord points for Tree",
    },
  ],
}
`;

exports[`user subcommand Return correct reply when user is not in database 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Points",
          "value": "OldUser has 0 points",
        },
        {
          "name": "Rank",
          "value": "OldUser currently has no rank in the leaderboard",
        },
      ],
      "title": "TOP Discord points for OldUser",
    },
  ],
}
`;
